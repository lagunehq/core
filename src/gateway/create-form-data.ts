// tslint:disable-next-line no-import-side-effect
import 'isomorphic-form-data';

export const isArray = (x: any): x is any[] =>
  Object.prototype.toString.call(x) === '[object Array]';
export const isObject = (x: any): x is { [key: string]: any } =>
  Object.prototype.toString.call(x) === '[object Object]';

export function createFormData(
  encodable: any,
  result = new FormData(),
  parentKey = '',
) {
  if (isArray(encodable)) {
    encodable.forEach(value => {
      createFormData(value, result, `${parentKey}[]`);
    });

    return result;
  }

  if (isObject(encodable)) {
    Object.entries<any>(encodable).forEach(([key, value]) =>
      createFormData(value, result, parentKey ? `${parentKey}[${key}]` : key),
    );

    return result;
  }

  result.append(parentKey, encodable as any);

  return result;
}
