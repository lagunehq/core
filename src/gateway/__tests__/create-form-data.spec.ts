import { createFormData, isArray, isObject } from '../create-form-data';

test('typed isArray', () => {
  const r1 = isArray([]);
  const r2 = isArray([1, 2]);
  const r3 = isArray({});
  const r4 = isArray(null);

  expect(r1).toBe(true);
  expect(r2).toBe(true);
  expect(r3).toBe(false);
  expect(r4).toBe(false);
});

test('typed isObject', () => {
  const r1 = isObject(null);
  const r2 = isObject([1, 2]);
  const r3 = isObject({});
  const r4 = isObject({ a: 'a' });

  expect(r1).toBe(false);
  expect(r2).toBe(false);
  expect(r3).toBe(true);
  expect(r4).toBe(true);
});

test('flat value', () => {
  const result = createFormData({
    apple: 'ringo',
    orange: 'mikan',
    grapes: 'budo',
  });

  const manual = new FormData();
  manual.append('apple', 'ringo');
  manual.append('orange', 'mikan');
  manual.append('grapes', 'budo');

  expect(result).toEqual(manual);
});

test('array', () => {
  const result = createFormData({
    animals: ['lion', 'giraffe', 'elephant'],
  });

  const manual = new FormData();
  manual.append('animals[]', 'lion');
  manual.append('animals[]', 'giraffe');
  manual.append('animals[]', 'elephant');

  expect(result).toEqual(manual);
});

test('nested object', () => {
  const result = createFormData({
    a: 'string',
    b: 123,
    c: [1, 2, 3],
    e: {
      e1: 'string',
      e2: {
        e21: {
          e211: 'string',
        },
        e22: [1, 2, 3],
      },
    },
  });

  const manual = new FormData();
  manual.append('a', 'string');
  manual.append('b', '123');
  manual.append('c[]', '1');
  manual.append('c[]', '2');
  manual.append('c[]', '3');
  manual.append('e[e1]', 'string');
  manual.append('e[e2][e21][e211]', 'string');
  manual.append('e[e2][e22][]', '1');
  manual.append('e[e2][e22][]', '2');
  manual.append('e[e2][e22][]', '3');

  expect(result).toMatchSnapshot();
  expect(result).toEqual(manual);
});
