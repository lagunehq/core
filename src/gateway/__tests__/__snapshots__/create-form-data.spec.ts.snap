// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nested object 1`] = `
FormData {
  "_boundary": "--------------------------623332509302484328564367",
  "_currentStream": null,
  "_insideLoop": false,
  "_overheadLength": 1058,
  "_pendingNext": false,
  "_released": false,
  "_streams": Array [
    "----------------------------623332509302484328564367
Content-Disposition: form-data; name=\\"a\\"

",
    "string",
    [Function],
    "----------------------------623332509302484328564367
Content-Disposition: form-data; name=\\"b\\"

",
    "123",
    [Function],
    "----------------------------623332509302484328564367
Content-Disposition: form-data; name=\\"c[]\\"

",
    "1",
    [Function],
    "----------------------------623332509302484328564367
Content-Disposition: form-data; name=\\"c[]\\"

",
    "2",
    [Function],
    "----------------------------623332509302484328564367
Content-Disposition: form-data; name=\\"c[]\\"

",
    "3",
    [Function],
    "----------------------------623332509302484328564367
Content-Disposition: form-data; name=\\"e[e1]\\"

",
    "string",
    [Function],
    "----------------------------623332509302484328564367
Content-Disposition: form-data; name=\\"e[e2][e21][e211]\\"

",
    "string",
    [Function],
    "----------------------------623332509302484328564367
Content-Disposition: form-data; name=\\"e[e2][e22][]\\"

",
    "1",
    [Function],
    "----------------------------623332509302484328564367
Content-Disposition: form-data; name=\\"e[e2][e22][]\\"

",
    "2",
    [Function],
    "----------------------------623332509302484328564367
Content-Disposition: form-data; name=\\"e[e2][e22][]\\"

",
    "3",
    [Function],
  ],
  "_valueLength": 27,
  "_valuesToMeasure": Array [],
  "dataSize": 0,
  "maxDataSize": 2097152,
  "pauseStreams": true,
  "readable": true,
  "writable": false,
}
`;
